cmd/ipfs/daemon.go

```diff
@@ -230,15 +233,31 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
 	// running in an uninitialized state.
 	initialize, _ := req.Options[initOptionKwd].(bool)
 	if initialize {
+		cfgLocation, _ := req.Options[initConfigOptionKwd].(string)
+		profiles, _ := req.Options[initProfileOptionKwd].(string)
 
-		cfg := cctx.ConfigRoot
-		if !fsrepo.IsInitialized(cfg) {
-			profiles, _ := req.Options[initProfileOptionKwd].(string)
+		if cfgLocation != "" {
```

Comment:
We shouldn't _re-initialize_ if the config is specified. We need to bypass all this logic if `fsrepo.IsInitialized` returns true.

Reply:
My understanding of the intent of `init`/`--init`, is that we'll be initalizing a new repo.
In this section, if the location of an existing config is specified, we parse it from that location, and write it out to the new repo/config location (`$IPFS_PATH`).
Basically using it as user specified template.

I may be wrong here, but I believe we don't reinitialize (we error when that's attempted).
We check the init target for existence, not the existing config location that was passed to us (through `--init-config-file`).
We populate the repo target config file, from the value passed in to us. Effectively copying it, then proceeding with daemon startup.

`fsrepo.IsInitialized()` is checked:
When the config location is not specified, we check in the else clause here:
https://github.com/ipfs/go-ipfs/blob/f1f726cd70ea7cf1c62d0c20a2a8613fb889b89e/cmd/ipfs/daemon.go#L255
When it is specified, we hit the check in doInit() here:
https://github.com/ipfs/go-ipfs/blob/f1f726cd70ea7cf1c62d0c20a2a8613fb889b89e/cmd/ipfs/init.go#L158

afterwards we continue the daemon startup process.

I believe we're meeting this expected behaviour.

Reply:
The intent of `--init` is to initialize if not already initialized.

> Initialize ipfs with default settings if not already initialized.

Reply:
I believe the current patch matches that. The only exception is that we use the settings from the specified config instead of the default config settings.

Reply:
The difference is that we return an _error_ when we do it in `doInit` (aborting the daemon).

Reply:
I've made the patch a little bit more sane.
If I'm following correctly, the intent is to abort the daemon if the config exists.
If we're told to initialize during `daemon` when a repo is already initalized, doInit() will return an already-initialized error to us, which halts daemon startup.
`init` does the same.

![image](https://user-images.githubusercontent.com/13862850/60994947-83ad5800-a31f-11e9-9a7d-c8a54423ee30.png)

Reply:
>  If I'm following correctly, the intent is to abort daemon startup if the repo already exists.

The intent is to not re-initialize if the repo already exists. The "--init" flag is designed to auto-initialize go-ipfs on first start _when necessary_. The `--init-config` flag shouldn't change this.

Reply:
Oh I see what's going on now. I was confusing/conflating `--init` and `init` behaviour.
In the latest patch, `--init` won't abort the daemon if the repo is already initialized. 
But `init` will error if the repo is already initalized.
There's also a commit that adds a warning for this but I don't know how useful that is. (droppable)

Reply:
>  There's also a commit that adds a warning for this but I don't know how useful that is. (droppable)

There should be no warning/error. Running `ipfs daemon --init` with an initialized repo is normal behavior. I.e., I'd expect a decent system service to start IPFS with:

```
ipfs daemon --migrate --init
```

That way, the user can just install IPFS and start it, no setup required.

Reply:
Gotcha. 
Branch has been reset to the commit without the warning.